8주차 과제 - Spring REST Docs 적용, Docker로 빌드 및 배포

jdalma opened the pull request:
벌써 마지막 8주차 과제이네요 ㅎㅎ
한 주간 잘 부탁드립니다 !!

***

저번 6주차 과제에서 `특정 도메인 서비스를 위한 한 가지의 기능만 가진 인터페이스`에 대한 이야기를 해주시다가 과제가 끝났는데 이번 과제에서 추가 과제로 피드포워드를 받을 수 있을까요 ㅎㅎ
[`a6f9c6b`](https://github.com/CodeSoom/spring-week8-assignment-1/pull/67/commits/a6f9c6bbc09e9fceb438d031adad5a8db2bea122)와 같이 `User`도메인 기능별 인터페이스를 분리 해 놓았습니다.
이 인터페이스들을 **잘 사용한다**라는 말씀이 어떤 의미일지 궁금하네요

jdalma committed:
주석 작성

jdalma committed:
User 기능별 인터페이스 추가

jdalma committed:
Spring REST Docs, asciidoctor 추가

jdalma committed:
asciidoctor depnedsOn(test) 추가

jdalma committed:
ProductController 상품 조회 , 생성에 대한  asciidoc 생성

ssisksl77 commented:
좋습니다!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -17,7 +21,10 @@
 @Service
 @Transactional
-public class UserService {
+public class UserService implements UserDelete,
+                                    UserFind,
+                                    UserUpdate,
+                                    UserRegister {
```
UserService가 결국에는 모든 인터페이스를 구현하는군요! 

이렇게 된다면 유지보수할 때, 삭제, 조회, 수정, 추가 기능이 한 곳에 모여있게 됩니다!

그리고 사용하는 코드들도 `UserService` 를 사용하게된다면 4개의 모든 기능을 가진 객체를 쓰게되죠! 

우선 `UserService` 또한 기능을 나눠보는 것이 좋을 것 같아요!

jdalma replied:
컨트롤러와 서비스 사이에 인터페이스를 두니 서비스의 구현체 확장이 유연할 것 같네요!!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/application/userInterface/UserRegister.java:
```diff
@@ -0,0 +1,14 @@
+package com.codesoom.assignment.application.userInterface;
```
패키지에 -s 가 붙은 곳이 있고 아닌 곳이 있군요! 

네이밍 규칙은 팀마다 다르겠지만 일관성이 있는 것이 좋습니다!

jdalma committed:
mariadb 2.7.5 추가

jdalma committed:
mariadb 계정 정보 추가

jdalma committed:
패키지 네이밍 수정

jdalma committed:
User 기능별 인터페이스 네이밍 수정

jdalma committed:
UserInterface 패키지 이름 수정

jdalma committed:
UserService 기능별 분리

jdalma committed:
인스턴스 인터페이스로 변경

jdalma committed:
User 테스트 코드 수정

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/controller/UserController.java:
```diff
@@ -0,0 +1,95 @@
+@RequestMapping("/users")
+public class UserController {
+
+    private final UserRegisterInterface userRegisterService;
+    private final UserUpdateInterface userUpdateService;
+    private final UserDeleteInterface userDeleteService;
```
이제 삭제기능을 수정한다고하면 `UserDeleteInterface` 기능만을 손보면 됩니다. 

하나의 책임만을 하게 되었다고 할 수 있죠

좀 더 생각해보면 `UserController` 는 과연 하나의 일을 하는 걸까요?

이것에 정답은 없습니다. 하지만 왠지 회원가입, 수정, 삭제를 개선했을 때, 테스트하는 코드가 한 곳에 뭉쳐있다는 생각이 드는군요!

물론 지금도 아주 좋습니다. 하지만 경험삼아 한번 Controller도 나눠보면 좀 더 책임이 나뉘어질 것 같은 생각이 드는군요!

jdalma replied:
말씀해주신대로 진행하면서 이렇게 까지 분리할 수 있구나 싶었습니다 ㅎㅎ
책임을 세심히 나누다보니 테스트 코드 작성할 때나 수정할 때도 편하다는 생각이 들었습니다

한 가지 궁금한 것이
```java
  @Override
  public User deleteUser(Long id) {
      User user = userFindService.findUser(id);
      user.destroy();
      return user;
  }
```
위와 같은 상황에서 사용자를 삭제하기 위해 `FindService`와 `DeleteService`가 **협력을 하고있다** 라고 생각이 드는데요
**협력 지점**을 찾는 것이 고민이 들었습니다.

1. 컨트롤러에서 `deleteUser()`를 호출할 때 `FindService.findUser(id)`결과를 담아준다.
    - `deleteUser(Long id , User findUser)`
2. DeleteService에서 UserRepository를 주입받아 **직접 사용자를 조회**한다.
3. 지금과 같이 Service끼리 협력한다.

위와 관련해서 다른 분들과 이야기를 나눈 적이 있는데 결론은 **수정이 최소로 일어날 것 같은 지점을 찾는것**이 였는데요
지금은 Service끼리 협력하지만, 만약 `FindService.findUser(id)`가 수정이 빈번히 일어날 지점 같으면 Repository를 직접 주입받아 사용하는 것이 좋을 것 같다라는 이야기로 끝났습니다.

영환님은 어떻게 결정하시는지 궁금하네요ㅎㅎ

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/controller/UserController.java:
```diff
@@ -0,0 +1,95 @@
+        userDeleteService.deleteUser(id);
+    }
+
+    private UserResultData getUserResultData(User user) {
+        if (user == null) {
+            return null;
+        }
+
+        return UserResultData.builder()
+                .id(user.getId())
+                .email(user.getEmail())
+                .name(user.getName())
+                .build();
+    }
```
이것은 `UserResultData` 안으로 이동시키는 것이 좀 더 바람직해보입니다!

jdalma committed:
profiles mariadb 추가

jdalma committed:
UserResultData of() 메소드 추가

jdalma committed:
h2 db 정보 수정

jdalma committed:
Role 저장 기능 및 테스트 추가

jdalma committed:
UserController 기능별 분리 및 테스트 코드 추가

jdalma commented:
계층형 테스트 코드는 내부 클래스 마다 어노테이션을 반복적으로 작성해줘야 해서 `RestDocs`와 관련된 테스트 코드는 `given-when-then`으로 작성하였습니다.

jdalma committed:
UserController docs 식별자 수정

jdalma committed:
User 추가

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/controller/user/UserRegisterController.java:
```diff
@@ -0,0 +1,38 @@
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    UserResultData create(@RequestBody @Valid UserRegistrationData registrationData) {
+        return UserResultData.of(userRegisterService.registerUser(registrationData));
```
질문이 있습니다. userRegisterService.registerUser 함수에서 null을 리턴하면 
HttpStatus.CREATED 상태와 함께 null을 리턴하나요?

jdalma replied:
UserRegisterService에서 `null`을 반환할 케이스가 없는 것 같아서 `mock`을 사용해서 테스트 해봤습니다.

> HttpStatus.CREATED 상태와 함께 null을 리턴하나요?

```
Status = 201
Error message = null
Headers = []
Content type = null
Body = 
```
위와 같이 바디는 공백으로 반환됐습니다 

[`35391f0`](https://github.com/CodeSoom/spring-week8-assignment-1/pull/67/commits/35391f0734570a351ec1f5c1dd992403e35f61d5)

ssisksl77 replied:
201 은 CREATED를 말하는데 그렇다면 요청된 정보를 생성한 것으로 의미하나요?

jdalma replied:
네 지금은 그런 상황인 것 같네요..
DB에 문제가 있는 상황을 가정하고 컨트롤러에서 null을 확인해서 예외를 던져 어드바이스에서 잡아줘야 할 것 같습니다
컨트롤러가 서비스를 주입받으니 서로 독립적이게 생각했어야 했는데 서비스에서 발생할 예외 상황만 생각하고 컨트롤러 입장에서의 예외 상황을 생각하지 못 했네요..

이상동작을 예상하기가 어렵네요 ㅎㅎ
영환님은 이 정도의 이상동작을 예상하시나요? 아니면 이런 상황이 발생하면 그때 테스트 코드를 추가하시는지 궁금하네요

ssisksl77 replied:
물론 우리는 최선을 다해 개발을 하고 에지케이스를 상상하며 개발을 하지만
무언가를 발견하면 그때그때 테스트를 추가하면서 해결한 문제들을 쌓아갑니다!

지금처럼 코드리뷰가 서로에게 의지되겠죠? ㅎㅎㅎ

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/application/user/UserUpdateService.java:
```diff
@@ -0,0 +1,45 @@
+     * @return 수정된 사용자
+     */
+    @Override
+    public User updateUser(Long id, UserModificationData modificationData,
+                           Long userId) throws AccessDeniedException {
```
여기서 id와 userId의 차이는 무엇인가요?!

jdalma replied:
`PATCH /users/{id}` 요청이 들어온다면 `경로 변수에 해당하는 사용자의 id`와 `토큰에 들어 있는 id`가 다르다면 예외를 주기위한 파라미터 입니다

컨트롤러에서 바로 예외를 던져서 처리하는게 나을까 라는 생각이 드네요 ㅎㅎ

ssisksl77 replied:
어쩌면 문서에 설명을 해놓는 것이 좋을 것 같습니다. `userId` 는 문서에 안보였었거든요 ㅠㅜㅠㅜ

다른 의견은 `userId` 가  `UserModificationData` 안에 들어가야하는 것이 아닌가 라는 생각도 들어요!

이에 대해 어떻게 생각하시는지 궁금합니다!

jdalma replied:
아 주석에 작성을 안해놓았었네요 ㅠㅠ

말씀하신대로 `UserModificationData `에 들어가는 것도 좋은 방법인 것 같습니다!!
`userId` 파라미터를 추가하면서 다른 방법이 없을까 고민했었는데.. `userId`필드를 DTO에 추가해 보았습니다 ㅎㅎ
[`540698d`](https://github.com/CodeSoom/spring-week8-assignment-1/pull/67/commits/540698dd938dd55d1b2f924c41fdcae9de52ecb9)

jdalma committed:
SecurityConfig 생성자 주입으로 수정

jdalma committed:
private 기본 생성자 추가

jdalma committed:
UserRegisterController 엣지 케이스 테스트 추가

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controller/user/UserDeleteControllerTest.java:
```diff
@@ -0,0 +1,67 @@
+    @Autowired
+    private RoleRegisterInterface roleRegisterService;
+
+    private UserRegistrationData newUserRegistrationData(String text){
+        return new UserRegistrationData(text , text ,text);
+    }
```
이것은 독자를 더욱 헷갈리게 하는 것 같습니다. 

인자 이름 `text` 가 실제 인자의 이름을 가질 수 없고 범용으로만 쓰이는 이유가 여러 개의 의미를 통합하기 때문인 것으로 보입니다!

그렇기에 순산 이 메소드만 봐서는 무엇을 하는 것인지 인지하지 못하고, 다시 한 번 `UserRegistrationData` 생성자가 어떻게 구현되어 있는지 보게만듭니다!

jdalma replied:
[`cc6f1cc`](https://github.com/CodeSoom/spring-week8-assignment-1/pull/67/commits/cc6f1ccae17bf20b8836b96b5fa77fc924c0b51d) 빌더를 사용해보았습니다!!

jdalma committed:
테스트 코드 DTO 생성 메소드 수정

jdalma committed:
UserUpdate DTO userId 필드, isDifferentUser 메소드 추가

jdalma committed:
UserUpdate 메소드 userId파라미터 제거로 인한 수정

jdalma committed:
ResourceCreateException 커스텀 예외 추가

Advice ResourceCreateException 핸들러 추가
사용자 생성 mock 테스트 코드 수정
UserRegisterController ResourceCreateException 주석 추가

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controller/user/UserDeleteControllerTest.java:
```diff
@@ -38,7 +38,11 @@ class UserDeleteControllerTest {
 
     private UserRegistrationData newUserRegistrationData(String text){
-        return new UserRegistrationData(text , text ,text);
+        return UserRegistrationData.builder()
+                                    .email(text)
+                                    .name(text)
+                                    .password(text)
+                                    .build();
```
오 좋습니다!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/controller/user/UserRegisterController.java:
```diff
@@ -28,11 +30,16 @@ public UserRegisterController(UserRegisterInterface userRegisterService) {
     UserResultData create(@RequestBody @Valid UserRegistrationData registrationData) {
-        return UserResultData.of(userRegisterService.registerUser(registrationData));
+        User user = userRegisterService.registerUser(registrationData);
+        if(user == null){
+            throw new ResourceCreateException("사용자 수정 메소드의 반환 값이 존재하지 않습니다.");
+        }
```
좋습니다! 

user가 null을 리턴하는 경우는 일반적으로 무엇일까요?!

ssisksl77 replied:
메시지만으로는 어떤 문제가 있는지 명확하게 알 수가 없을 것 같아요!

물론 무슨 일이 일어났는지 디버그를 할 수 있으면 좋겠지만 상황이 여의치 않다면 메시지만으로 어느정도 짐작이 가능할 수 있으면 좋을 것 같아요!

jdalma replied:
> user가 null을 리턴하는 경우는 일반적으로 무엇일까요?!

DTO 검증도 이뤄지고 있고 `findUser`에서 사용자 존재도 확인하고 있으니 DB에 문제가 있는 상황일 것 같습니다. 

> 메시지만으로 어느정도 짐작이 가능할 수 있으면 좋을 것 같아요!

`throw new ResourceCreateException("사용자 생성에 실패하였습니다.[registerUser에서 null이 반환되었습니다.]");` 이렇게 수정 해보았습니다 ㅎㅎ

jdalma committed:
사용자 생성 ResourceCreateException 메시지 수정

jdalma committed:
UserRegisterControllerMock MockMvc UTF-8 필터 추가

jdalma commented:
말씀해주신대로 컨트롤러와 서비스를 기능별로 분리하면서 이렇게 까지 분리할 수 있구나 싶었습니다 ㅎㅎ
책임을 세심히 나누다보니 테스트 코드 작성할 때나 수정할 때도 편하다는 생각이 들었습니다

한 가지 궁금한 것이
```java
  @Override
  public User deleteUser(Long id) {
      User user = userFindService.findUser(id);
      user.destroy();
      return user;
  }
```
위와 같은 상황에서 사용자를 삭제하기 위해 `FindService`와 `DeleteService`가 **협력을 하고있다** 라고 생각이 드는데요
**협력 지점**을 찾는 것이 고민이 들었습니다.

1. 컨트롤러에서 `deleteUser()`를 호출할 때 `FindService.findUser(id)`결과를 담아준다.
    - `deleteUser(Long id , User findUser)`
2. DeleteService에서 UserRepository를 주입받아 **직접 사용자를 조회**한다.
3. 지금과 같이 Service끼리 협력한다.

위와 관련해서 다른 분들과 이야기를 나눈 적이 있는데 결론은 **수정이 최소로 일어날 것 같은 지점을 찾는것**이 였는데요
지금은 Service끼리 협력하지만, 만약 `FindService.findUser(id)`가 수정이 빈번히 일어날 지점 같으면 Repository를 직접 주입받아 사용하는 것이 좋을 것 같다라는 이야기로 끝났습니다.

영환님은 어떻게 결정하시는지 궁금하네요ㅎㅎ

jdalma commented:
마지막 회고이네요 ㅎㅎ
그 동안 감사했습니다!!
[회고](https://jdalma.github.io/docs/retrospective/2022y10m/#10%EC%9B%94-%EC%B2%AB-%EC%A7%B8%EC%A3%BC-%ED%9A%8C%EA%B3%A0-%EC%BD%94%EB%93%9C%EC%88%A8-%EB%81%9D)

ssisksl77 commented:
> 말씀해주신대로 컨트롤러와 서비스를 기능별로 분리하면서 이렇게 까지 분리할 수 있구나 싶었습니다 ㅎㅎ 책임을 세심히 나누다보니 테스트 코드 작성할 때나 수정할 때도 편하다는 생각이 들었습니다
> 
> 한 가지 궁금한 것이
> 
> ```java
>   @Override
>   public User deleteUser(Long id) {
>       User user = userFindService.findUser(id);
>       user.destroy();
>       return user;
>   }
> ```
> 
> 위와 같은 상황에서 사용자를 삭제하기 위해 `FindService`와 `DeleteService`가 **협력을 하고있다** 라고 생각이 드는데요 **협력 지점**을 찾는 것이 고민이 들었습니다.
> 
> 1. 컨트롤러에서 `deleteUser()`를 호출할 때 `FindService.findUser(id)`결과를 담아준다.
>    
>    * `deleteUser(Long id , User findUser)`
> 2. DeleteService에서 UserRepository를 주입받아 **직접 사용자를 조회**한다.
> 3. 지금과 같이 Service끼리 협력한다.
> 
> 위와 관련해서 다른 분들과 이야기를 나눈 적이 있는데 결론은 **수정이 최소로 일어날 것 같은 지점을 찾는것**이 였는데요 지금은 Service끼리 협력하지만, 만약 `FindService.findUser(id)`가 수정이 빈번히 일어날 지점 같으면 Repository를 직접 주입받아 사용하는 것이 좋을 것 같다라는 이야기로 끝났습니다.
> 
> 영환님은 어떻게 결정하시는지 궁금하네요ㅎㅎ

`FindService.findUser(id)` 구현의 수정이 빈번히 일어날 지점이라는 말씀으로 이해하겠습니다.

1. 수정이 모두 일어나야 한다면 좋은 것이다.
2. 만약 `findUser(id)` 의 구현이 어디는 바뀌고 어디는 바뀌지 않아야 한다면 스프링의 장점인 DI를 이용하여 `FindService` 구현을 또한 분리한다.

위처럼 생각할 것 같습니다.
